backends/platform/3ds/osystem-graphics.cpp

```diff
@@ -124,27 +172,77 @@ bool OSystem_3DS::getFeatureState(OSystem::Feature f) {
 	}
 }
 
-void OSystem_3DS::initSize(uint width, uint height,
-                                   const Graphics::PixelFormat *format) {
-	debug("3ds initsize w:%d h:%d", width, height);
-	int oldScreen = config.screen;
-	loadConfig();
-	if (config.screen != oldScreen) {
-		_screenChangeId++;
+GraphicsModeID OSystem_3DS::chooseMode(Graphics::PixelFormat *format) {
+	if (format->bytesPerPixel > 2) {
+//		if (format->aBits() > 0) {
+			return RGBA8;
+//		} else {
+//			return RGB8;
+//		}
+	} else if (format->bytesPerPixel > 1) {
+		if (format->gBits() > 5) {
+			return RGB565;
+		} else if (format->aBits() == 0) {
+			return RGB555;
+		} else {
+			return RGB5A1;
+		}
 	}
 
+	return CLUT8;
+}
+
+bool OSystem_3DS::setGraphicsMode(GraphicsModeID modeID) {
+	switch (modeID) {
+	case RGBA8:
+	case RGB565:
+	case RGB8:
+	case RGB555:
+	case RGB5A1:
+	case CLUT8:
+		_gfxState.gfxMode = gfxModes[modeID];
+		return true;
+	default:
+		return false;
+	}
+}
+
+void OSystem_3DS::initSize(uint width, uint height,
+                           const Graphics::PixelFormat *format) {
+	debug("3ds initsize w:%d h:%d", width, height);
 	_gameWidth = width;
 	_gameHeight = height;
-	_gameTopTexture.create(width, height, _pfGameTexture);
-	_overlay.create(400, 320, _pfGameTexture);
 	_magCenterX = _magWidth / 2;
 	_magCenterY = _magHeight / 2;
 
-	if (format) {
+	_oldPfGame = _pfGame;
+	if (!format) {
+		if (_pfGame != Graphics::PixelFormat::createFormatCLUT8()) {
+			_pfGame = Graphics::PixelFormat::createFormatCLUT8();
+		}
+	} else {
 		debug("pixelformat: %d %d %d %d %d", format->bytesPerPixel, format->rBits(), format->gBits(), format->bBits(), format->aBits());
 		_pfGame = *format;
 	}
 
+	/* If the current graphics mode does not fit with the pixel
+	 * format being requested, choose one that does and switch to it */
+	assert(_pfGame.bytesPerPixel > 0);
+	if (_pfGame != _oldPfGame) {
+		assert(_transactionState == kTransactionActive);
+		_gfxState.gfxModeID = chooseMode(&_pfGame);
+		_transactionDetails.formatChanged = true;
+	}
+
+	int oldScreen = config.screen;
+	loadConfig();
+	if (config.screen != oldScreen) {
+		_screenChangeId++;
+	}
+
+	_gameTopTexture.create(width, height, _gfxState.gfxMode);
+	_overlay.create(400, 320, _gfxState.gfxMode);
```

Comment:
I'm not sure I understand, why use the same pixel format for the game and the overlay? Wouldn't it be simpler and less problematic to always use RGBA for the overlay (that way the overlay always has alpha, removing the need for the _screenGrab surface)?

Reply:
### Short explanation:
Because the screen render targets can only accept a single "in" format, both the game and the overlay have to share the same format.

### Long Explanation:
As the 3DS backend stands currently, all surfaces are converted to RGBA8 before being passed on to the screen render targets. Additionally, the current backend uses Ctrulib's default graphics initialization, which initializes the LCD framebuffers to RGB8 (technically it's BGR8 in the 3DS's case but that's neither here nor there). In effect, games that do not already run in RGBA8, ARGB8, or whatever the backend most prefers (which in this case is also RGBA8) are being converted to two other formats each and every frame. In the case of a game like Riven, which has a game format of RGB565 (2 bytes), each frame is converted to RGBA8 (4 bytes) for the screen render targets, which is then converted into RGB8 (3 bytes) for the framebuffers. This is what causes huge slowdowns.

Unfortunately, simply changing the framebuffers to use RGBA8 as well does not solve the problem. What I found does solve the problem is to minimize instances of format conversions. This pull request does that by altering the format used throughout (surfaces, screen render targets, framebuffers, displays) to match or closely match the format used by whatever game is about to be played. Because render targets can only accept one "in" format, however, it is necessary to keep all surfaces (including the overlay and cursor) the same format.

Reply:
> Because the screen render targets can only accept a single "in" format, both the game and the overlay have to share the same format.

Are you sure though? See this branch here I've just updated: https://github.com/bgK/scummvm/tree/3ds-565

There the game texture is either RGBA or RGB565 depending on the game surface pixel format. The other surfaces are always RGBA. It seems to have similar performance characteristics when compared to yours.

Reply:
Well dang (in a good way), you're right. I'll get to work on changing the PR.

Also, I think the the keymapper commit caused a bug; after launching a game, the first touchscreen input doesn't work, causing each subsequent touchscreen input to execute the one that came before it. I thought it might be something on my branch causing it, but it's in the master branch as well.

Reply:
Thanks, I'll look into the keymapper bug.

Reply:
I've finished reworking the PR. Unless you have any other questions, it should be ready for merging.